ILM: Make all the shrink action steps retryable

andreidan opened the pull request:
This aims at making the shrink action retryable. Every step is
retryable, but in order to provide an experience where ILM tries
to achieve a successful shrink even when the target node goes
missing permanently or the shrunk index cannot recover, this also
introduces a retryable shrink cycle within the shrink action.

The shrink action will generate a unique index name that'll be the
shrunk index name. The generated index name is stored in the lifecycle
state.

If the shrink action ends up waiting for the source shards to
collocate or for the shrunk index to recover for more than the configured
`LIFECYCLE_STEP_WAIT_TIME_THRESHOLD` setting, it will move back
to clean up the attempted (and failed) shrunk index and will retry
generating a new index name and attempting to shrink the source
to the newly generated index name.

Relates to #48183

andreidan committed:
ILM: Make all the shrink action steps retryable

This aims at making the shrink action retryable. Every step is
retryably, but in order to provide an experience where ILM tries
to achieve a successful shrink even when the target node goes
missing permanently or the shrunk index cannot recover, this also
introduces a retryable shrink cycle within the shrink action.

The shrink action will generate a unique index name that'll be the
shrunk index name. The generated index name is stored in the lifecycle
state.

If the shrink action ends up waiting for the source shards to
colocate or for the shrunk index to recover for more than the configured
`LIFECYCLE_STEP_WAIT_TIME_THRESHOLD` setting, it will move back
to cleanup the attempted (and failed) shrunk index and will retry
generating a new index name and attempting to shrink the source
to the newly generated index name.

andreidan committed:
Cleanup shrink index step succeeds if the index doesn't exist

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retryable-shrink-action

andreidan committed:
Remove accidental invariant non-null nextStepKey

We were setting the nextStepKey regardless if the underlying step
execution was successful or not. We were also asserting we have a
`nextStepKey` even if the step failed (and moved into the `ERROR` step).

I don't think this is correct as from the `ERROR` step the only place
ILM can move is back to the `failedStep` (if it was retryable).

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:
```diff
@@ -47,6 +48,13 @@
+    // This setting configures how much time since step_time should ILM wait for a condition to be met. After the threshold wait time has
+    // elapsed ILM will likely stop waiting and go to the next step.
+    // Also see {@link org.elasticsearch.xpack.core.ilm.ClusterStateWaitUntilThresholdStep}
+    public static final Setting<TimeValue> LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING =
```
Would we want to document this?

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retryable-shrink-action

dakrone commented on x-pack/plugin/ilm/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/ilm/CCRIndexLifecycleIT.java:
```diff
@@ -650,20 +653,6 @@ private void putShrinkOnlyPolicy(RestClient client, String policyName) throws IO
                     }
                     builder.endObject();
-
-                    // Sometimes throw in an extraneous unfollow just to check it doesn't break anything
-                    if (randomBoolean()) {
```
Why was this removed?

andreidan replied:
Ah, I think I was suspecting this could cause some flakiness, but I believe it might've been due to dev issues as I was working through this feature (saving the execution state - namely the shrink index name being present in the execution state when we reach the cold pohase). I'll add this back. Thanks!

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:
```diff
@@ -47,6 +48,13 @@
+    // elapsed ILM will likely stop waiting and go to the next step.
+    // Also see {@link org.elasticsearch.xpack.core.ilm.ClusterStateWaitUntilThresholdStep}
+    public static final Setting<TimeValue> LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING =
+        Setting.positiveTimeSetting(LIFECYCLE_STEP_WAIT_TIME_THRESHOLD, TimeValue.timeValueHours(12), Setting.Property.Dynamic,
```
Should we enforce a minimum setting? It would be bad to be spinning forever because someone put 1 second as their wait time.

andreidan replied:
Good point. Shall we set it to one hour minimum? (this will make it impossible to integration test - https://github.com/elastic/elasticsearch/pull/70107/files/db8a13789e40a0b7f64263a0c88a257b21eb206c#diff-00728663bc72dd25e9832acf55fd6c4dd690a34d3c9ef6e5408464d81ff1f785R309 - however, maybe the unit tests we have around the `ClusterStateWaitUntilThresholdStep` are enough)

andreidan replied:
I managed to integration test the rewind even with a `1h` minimum value in https://github.com/elastic/elasticsearch/pull/70107/commits/73e363dec50e3773d90390f1e50582a55fa05b3b

dakrone reviewed (changes requested):
Thanks for working on this Andrei! I left some comments for an initial review.

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retryable-shrink-action

andreidan committed:
Make the operator precedence explicit

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java:
```diff
@@ -0,0 +1,154 @@
+        String generatedIndexName = generateValidIndexName(prefix, index.getName());
+        ActionRequestValidationException validationException = validateGeneratedIndexName(generatedIndexName, clusterState);
+        if (validationException != null) {
+            logger.warn("unable to generate a valid shrink index name as part of policy [{}] for index [{}] due to [{}]",
```
Ah yes, good catch

andreidan committed:
Append a random 4 char sequence to the index prefix

This changes the generated index name format to <prefix>-<uuid>-<indexname>, 
limitting the length of the uuid to a maximum of 4 chars.

andreidan committed:
Fix ILMMultiNodeIT

andreidan committed:
Rewind to setSingleNodeKey if the source index shards can't be colocated

andreidan committed:
Comment the shrink action steps

andreidan committed:
Add ShrinkIndexNameSupplier

andreidan committed:
Readd a random cold phase in the CCR lifecycle tests

andreidan committed:
Fix test

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ClusterStateWaitUntilThresholdStep.java:
```diff
@@ -0,0 +1,142 @@
+            // wonderful thing)
+            TimeValue retryThreshold = LifecycleSettings.LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING.get(idxMeta.getSettings());
+            LifecycleExecutionState lifecycleState = fromIndexMetadata(idxMeta);
+            if (waitedMoreThanThresholdLevel(retryThreshold, lifecycleState, Clock.systemUTC())) {
```
Oh, you're absolutely right - wall clocks are not trustworthy. I don't think `nanoTime` would be the answer and the reading invariant on that api is to be monotonically increasing, more so than accuracy.

I don't think this is too bad because if we get a reading where the time drift is hours away we'll retry the shrink cycle (so nothing truly bad happens, except a wasteful - and presumbaly eventually successful - shrink operation, with the note that we'll at least delete the abandoned shrunk index)

andreidan committed:
Unify usage of ShrinkIndexNameSupplier in the ShrinkAction

andreidan committed:
Enforce minimum value of 1h for the LIFECYCLE_STEP_WAIT_TIME_THRESHOLD setting

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retryable-shrink-action

andreidan commented:
@dakrone thanks for the review. I think I addressed all your comments

andreidan commented:
@jrodewig can you please help us with the docs section (I had an initial go at documenting the new behaviour of the shrink action)

jrodewig committed:
Refactor docs

jrodewig committed:
Reword + whitespace

jrodewig reviewed (commented):
Thank @andreidan. I pushed 0fa3ca2d08adc25054984511762fdf876af6e679 to refactor the docs a bit.

That commit moves the info about allocation and retrying steps to the bottom of the ILM shrink page. I feel that retrying steps should be relatively rare. Most users probably don't need to worry about it.

I also added some reference docs for `index.lifecycle.step.wait_time_threshold`. Please feel free to make further changes or reach out if you have questions. Thanks again!

andreidan commented on docs/reference/ilm/actions/ilm-shrink.asciidoc:
```diff
@@ -103,3 +91,25 @@ PUT _ilm/policy/my_policy
+index's shards to the appropriate nodes based on your allocation rules. These
+allocation steps can fail if:
+
+* A node is removed during the `shrink` action.
+* No node has enough disk space to host the source index's shards.
+* {es} cannot reallocate the shrunken index due to conflicting allocation rules.
```
@jrodewig these are some examples of the possible conditions that might not be fulfilled. Shall we rephrase this a bit to illustrate that? (as opposed to being the only possible issues)

jrodewig replied:
Thanks @andreidan. I rephrased as follows:
> These allocation steps can fail for a number of reasons, including:

andreidan commented on docs/reference/ilm/actions/ilm-shrink.asciidoc:
```diff
@@ -103,3 +91,25 @@ PUT _ilm/policy/my_policy
 }
 --------------------------------------------------
+
+[[ilm-shrink-retry-failed-allocation]]
+==== Retrying failed allocation steps
```
@jrodewig the threshold and retry-ability/rewind of the shrink action is based on waiting for some conditions that are not fulfilled until the threshold is breached/surpassed. Should we be focused more on the "wait until threshold and then rewind" as opposed to retrying a step?

jrodewig replied:
Good point. I renamed this "Shard allocation for shrink" as it could also be useful for people curious about how shrink allocates normally.

andreidan commented on docs/reference/settings/ilm-settings.asciidoc:
```diff
@@ -59,6 +59,15 @@ An index that was rolled over would normally match the full format,
+[[index-lifecycle-step-wait-time-threshold]]
+`index.lifecycle.step.wait_time_threshold`::
+(<<indices-update-settings,Dynamic>>, <<time-units,time value>>) 
+Time to wait while retrying a failed allocation step for the {ilm-init}
```
@jrodewig I'm not sure if we should use "while retrying" as we are not necessarily retrying steps, but rather waiting for conditions to be fulfilled (ie. source index shards be collocated, shrunk index shards be recovered/allocated)

jrodewig replied:
I reworded to:
> Time to wait for the cluster to resolve allocation issues during an ILM `shrink` action.

andreidan commented on docs/reference/settings/ilm-settings.asciidoc:
```diff
@@ -59,6 +59,15 @@ An index that was rolled over would normally match the full format,
+`index.lifecycle.step.wait_time_threshold`::
+(<<indices-update-settings,Dynamic>>, <<time-units,time value>>) 
+Time to wait while retrying a failed allocation step for the {ilm-init}
+<<ilm-shrink,`shrink`>> action. After this period, {ilm-init} re-attempts the
```
@jrodewig "re-attempts the previous step" is not entirely accurate as the shrink action rewinds back a few steps, depending on the situation. Maybe we shouldn't be so prescriptive here and say the `shrink` action (which is the only place where this is used now) "does the right thing"

jrodewig replied:
This is a good point. We don't really need to document the remedial action here anyway. That's already covered in the xref.

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java:
```diff
@@ -177,14 +180,33 @@ public boolean isSafeAction() {
             waitForNoFollowerStepKey);
         WaitForNoFollowersStep waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, readOnlyKey, client);
-        UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, setSingleNodeKey, client, readOnlySettings);
+        UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, cleanupShrinkIndexKey, client, readOnlySettings);
```
While we're changing this step, can you make this use `ReadOnlyStep` instead of the generic settings update one?

dakrone reviewed (changes requested):
Thanks Andrei, I left some more comments. I also played around with this manually and found an interesting thing. If we don't use a number of shards (using `max_primary_shard_size` instead), then if manually moving an index that is *past* a shrink step back to any point *before* the shrink, the cleanup step actually ends up deleting the index itself!

I think it'd be great if we could figure out a way to avoid this, maybe in the `CleanupShrinkIndexStep` we can check to see if the index name matches the pre-existing name in `lifecycleExecutionState.getShrinkIndexName()` then we skip doing the shrink? What do you think? (I know this is an edge case, I was just curious to see how it behaved)

jrodewig committed:
Reword based on feedback

jrodewig commented:
Thanks for the review and iteration @andreidan. I pushed a few more edits to the docs with 2ac08e741a0a7d8831d0d0fea0ebe9a0306a3f8c. Happy to address any other feedback you have!

jrodewig committed:
Minor edit

jrodewig committed:
whitespace

jrodewig committed:
whitespace

jrodewig committed:
missing word

jrodewig committed:
reword

jrodewig committed:
reword

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateUniqueIndexNameStep.java:
```diff
@@ -0,0 +1,154 @@
+        return prefix + randomUUID + "-" + indexName;
+    }
+
+    static String generateValidIndexSuffix(Supplier<String> randomGenerator) {
```
I pondered that too, but I still think the suffix is still a better name as it might not be valid if it'd be used as a prefix (ie. we don't make sure the generated UUID doesn't *start with* `_` or `-`)

dakrone replied:
Okay, that's fine with me :)

andreidan commented:
@jrodewig thanks so much for restructuring and rewording the docs

andreidan committed:
Skip the cleanup index step if the managed index is a shrunk index and the
source index does not exist anymore

andreidan commented:
> if manually moving an index that is past a shrink step back to any point before the shrink, the cleanup step actually ends up deleting the index itself

@dakrone ah interesting. So this happened because the alias/identity of the source index was swapped to the shrink index (and the source index was deleted in the process). I pushed https://github.com/elastic/elasticsearch/pull/70107/commits/4b091e683b71ce57c8de63b90a8d3da014aa4868 to skip the cleanup index step if the managed index is a shrunk index and the source index does not exist anymore

andreidan committed:
Convert if check to assert

andreidan commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ExecuteStepsUpdateTask.java:
```diff
@@ -127,6 +124,10 @@ public ClusterState execute(final ClusterState currentState) throws IOException
+                    // some steps can decide to change the next step to execute after waiting for some time for the condition
+                    // to be met (eg. {@link LifecycleSettings#LIFECYCLE_STEP_WAIT_TIME_THRESHOLD_SETTING}, so it's important we
+                    // re-evaluate what the next step is after we evaluate the condition
+                    nextStepKey = currentStep.getNextStepKey();
```
Great catch! Removed `nextStepKey` from the logging statement as it's logged below lines 132+

andreidan committed:
Drop null `nextStepKey` from log statement

andreidan committed:
generateValidIndexSuffix lowercases input

andreidan committed:
Shrink action uses the readonly step and test fixes

elasticmachine committed:
Merge branch 'master' into ilm-retryable-shrink-action

andreidan committed:
Disable bwc tests as the explain API output is modified in this PR

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-retryable-shrink-action

dakrone reviewed (approved):
LGTM, thanks for iterating on this Andrei!
